### ISSUE TITLE
loss is NAN

### ISSUE THREAD

**Issue by missbook520** [ziwei-jiang/PGGAN-PyTorch#1]
Thank you very much for your work. When I was training my data set, I found that starting from the second epoch, the loss is nan. Can you help me?
![1648523029(1)](https://user-images.githubusercontent.com/48777650/160524704-393ca04a-c164-437c-aadb-3f0a0564be68.png)

**Comment by ziwei-jiang** [comment]
Hello, I think you might need to try different hyper-parameters such as batch size/ learning rates.
It also might be helpful to pre-process the data so that the variance in your data might be lower(similar to alignment for the face dataset). But other than that I'm not sure how this issue could be resolved.
Good luck!

**Comment by avr4L** [comment]
Hi, I had the same problem and could trace it to the class "Minibatch_std" in utils.py. 
Here is what happens: Lets say you have a dataset of 65 images and a batch_size of 8. For the first 8 iterations you get your batch with 8 samples and everything is fine. At the 9th iteration only one sample is left in the dataset, so you end up with a batch of shape [1, 3, img_size, img_size]. When you feed that batch into the discriminator it ends up in the Minibatch_std which tries to calculate "std = torch.std(x, dim=0)". This results in nan values since you try to calculate the standard deviation of a single number.
The quick fix I used was to add the following in the train.py script:
for i, samples in enumerate(databar):
____ if samples[0].shape[0] == 1:
________ continue
In case you have a batch with only one sample, skip the iteration

**Comment by ziwei-jiang** [comment]
> Hi, I had the same problem and could trace it to the class "Minibatch_std" in utils.py. Here is what happens: Lets say you have a dataset of 65 images and a batch_size of 8. For the first 8 iterations you get your batch with 8 samples and everything is fine. At the 9th iteration only one sample is left in the dataset, so you end up with a batch of shape [1, 3, img_size, img_size]. When you feed that batch into the discriminator it ends up in the Minibatch_std which tries to calculate "std = torch.std(x, dim=0)". This results in nan values since you try to calculate the standard deviation of a single number. The quick fix I used was to add the following in the train.py script: for i, samples in enumerate(databar): ____ if samples[0].shape[0] == 1: ________ continue In case you have a batch with only one sample, skip the iteration

Thank you very much for catching the problem! That's a good point. I'm thinking to add the "drop_last" option in the dataloader, so there won't be any instability due to small batch size during the training.
Thanks!

**Comment by avr4L** [comment]
> > Hi, I had the same problem and could trace it to the class "Minibatch_std" in utils.py. Here is what happens: Lets say you have a dataset of 65 images and a batch_size of 8. For the first 8 iterations you get your batch with 8 samples and everything is fine. At the 9th iteration only one sample is left in the dataset, so you end up with a batch of shape [1, 3, img_size, img_size]. When you feed that batch into the discriminator it ends up in the Minibatch_std which tries to calculate "std = torch.std(x, dim=0)". This results in nan values since you try to calculate the standard deviation of a single number. The quick fix I used was to add the following in the train.py script: for i, samples in enumerate(databar): ____ if samples[0].shape[0] == 1: ________ continue In case you have a batch with only one sample, skip the iteration
> 
> Thank you very much for catching the problem! That's a good point. I'm thinking to add the "drop_last" option in the dataloader, so there won't be any instability due to small batch size during the training. Thanks!

Thank you for your work, this is probably the most straight forward pytoch implementation of PGGAN. "drop_last" indeed seems to be the cleaner solution. Cheers mate.
